Adding support for Efinix Titanium Ti60 F225 Development Kit  BSP & Sapphire RISC-V SoC

manoj153 opened the pull request:
This PR adds support for the Titanium Ti60 F225 Development Kit from Efinix

The initial support for the board includes UART, PLIC, GPIO, CLINT and the RISC-V privilege core.

Relevant Links
Vendor: https://www.efinixinc.com/
Board : https://www.efinixinc.com/products-devkits-titaniumti60f225.html
Sapphire SoC: https://www.efinixinc.com/products-riscv.html

manoj153 committed:
dts: bindings: vendor-prefixes: Add efinix prefix

Add efinix manufacturer binding prefix.

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 committed:
dts: bindings: gpio: Add `efinix,sapphire-gpio`

A new gpio controller addition

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 committed:
dts: bindings: serial: Add `efinix,sapphire-uart0`

A new UART controller addition, interrupt is optional

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 committed:
dts: bindings: timer: Add `efinix,sapphire-timer0`

A new timer controller addition

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 committed:
dts: riscv: add a initial SoC dtsi for Efinix Sapphire SoC

Sapphire SoC from Efinix

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 committed:
drivers: gpio: gpio_efinix_sapphire: initial version

Add GPIO driver for Efinix Sapphire SoCs.

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 committed:
drivers: serial: add uart_efinix_sapphire initial version

Add UART driver for Efinix Sapphire SoCs

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 committed:
soc: riscv: Add initial support for Efinix Sapphire SoC

- It's a riscv privilege spec SoC

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

fkokosinski commented on drivers/serial/uart_efinix_sapphire.c:
```diff
@@ -0,0 +1,228 @@
+	return 0;
+}
+
+static int uart_efinix_sapphire_fifo_fill(const struct device *dev, const uint8_t *tx_data, int len)
```
Why are we ignoring the `len` argument? We should be writing here as many bytes from the `tx_data` buffer as we can fit in the FIFO.

fkokosinski commented on drivers/serial/uart_efinix_sapphire.c:
```diff
@@ -0,0 +1,228 @@
+	return 1;
+}
+
+static int uart_efinix_sapphire_fifo_read(const struct device *dev, uint8_t *rx_data, const int len)
```
Ditto - we're ignoring `len` here as well.

fkokosinski commented on soc/riscv/riscv-privilege/efinix-sapphire/linker.ld:
```diff
@@ -0,0 +1,3 @@
+# SPDX-License-Identifier: Apache-2.0
+
+#include <zephyr/arch/riscv/common/linker.ld>
```
Nitpick: add newline at the end of this file.

fkokosinski reviewed (dismissed):
Also: please add board documentation.

carlocaione commented on soc/riscv/riscv-privilege/efinix-sapphire/soc.h:
```diff
@@ -0,0 +1,47 @@
+
+static inline void sapphire_write8(uint32_t addr, uint8_t data)
+{
+	sys_write8(data, addr);
```
again, why this needs to be redefined?

carlocaione commented on soc/riscv/riscv-privilege/efinix-sapphire/soc.h:
```diff
@@ -0,0 +1,47 @@
+
+static inline void sapphire_write32(uint32_t addr, uint32_t data)
+{
+	sys_write32(data, addr);
```
ditto

carlocaione commented on soc/riscv/riscv-privilege/efinix-sapphire/soc.h:
```diff
@@ -0,0 +1,47 @@
+
+static inline uint32_t sapphire_read32(uint32_t addr)
+{
+	return sys_read32(addr);
```
ditto

carlocaione commented on drivers/gpio/gpio_efinix_sapphire.c:
```diff
@@ -0,0 +1,242 @@
+		return -ENOTSUP;
+	}
+
+	if ((flags & GPIO_OUTPUT) && (flags & GPIO_INPUT)) {
```
`flags & (GPIO_OUTPUT | GPIO_INPUT)` same for all the other occurrences

carlocaione commented on drivers/gpio/gpio_efinix_sapphire_priv.h:
```diff
@@ -0,0 +1,19 @@
+#ifndef ZEPHYR_DRIVERS_GPIO_EFINIX_SAPPHIRE_PRIV_H_
+#define ZEPHYR_DRIVERS_GPIO_EFINIX_SAPPHIRE_PRIV_H_
+
+#define BSP_GPIO_INPUT                  0x00
+#define BSP_GPIO_OUTPUT                 0x04
+#define BSP_GPIO_OUTPUT_ENABLE          0x08
+#define BSP_GPIO_INTERRUPT_RISE_ENABLE  0x20
+#define BSP_GPIO_INTERRUPT_FALL_ENABLE  0x24
+#define BSP_GPIO_INTERRUPT_HIGH_ENABLE  0x28
+#define BSP_GPIO_INTERRUPT_LOW_ENABLE   0x2c
```
use `BIT_MASK`

carlocaione commented on drivers/gpio/CMakeLists.txt:
```diff
@@ -81,3 +81,4 @@ if(CONFIG_GPIO_SC18IM704)
   zephyr_library_include_directories(${ZEPHYR_BASE}/drivers)
   zephyr_library_sources(gpio_sc18im704.c)
 endif()
+zephyr_library_sources_ifdef(CONFIG_GPIO_EFINIX_SAPPHIRE    gpio_efinix_sapphire.c)
```
move this to line 79

carlocaione commented on drivers/serial/uart_efinix_sapphire.c:
```diff
@@ -0,0 +1,227 @@
+static void uart_efinix_sapphire_poll_out(const struct device *dev, unsigned char c)
+{
+	/* uart_writeAvailability */
+	while ((sapphire_msb_read16(UART0_STATUS_REG_ADDR) & 0xFF) == 0) {
```
`sapphire_msb_read8` ?

carlocaione commented on drivers/serial/uart_efinix_sapphire.c:
```diff
@@ -0,0 +1,227 @@
+
+static int uart_efinix_sapphire_fifo_fill(const struct device *dev, const uint8_t *tx_data, int len)
+{
+	uint32_t fifo_size = (sapphire_read32(UART0_STATUS_REG_ADDR) >> 16) & 0xFF;
```
`sapphire_msb_read8` ?

carlocaione commented on drivers/serial/uart_efinix_sapphire.c:
```diff
@@ -0,0 +1,227 @@
+{
+	ARG_UNUSED(dev);
+
+#define UART0_SAMPLE_PER_BAUD 8
+#define UART0_PARITY	      0 /* Off */
+#define UART0_STOP	      0 /* 1 stop bit */
```
defines at the top

carlocaione commented on drivers/serial/uart_efinix_sapphire_priv.h:
```diff
@@ -0,0 +1,15 @@
+#ifndef __UART_EFINIX_SAPPHIRE_PRIV_H__
+#define __UART_EFINIX_SAPPHIRE_PRIV_H__
+
+#define BSP_UART_DATA           0x00
+#define BSP_UART_STATUS         0x04
+#define BSP_UART_CLOCK_DIVIDER  0x08
+#define BSP_UART_FRAME_CONFIG   0x0C
```
`BIT()`

manoj153 committed:
boards: riscv: Add initial support for titanium_ti60_f225

This FPGA board can be loaded with a RISC-V core and utilized
with Zephyr.

Product link:
https://www.efinixinc.com/products-devkits-titaniumti60f225.html

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 commented on dts/riscv/efinix/sapphire_soc.dtsi:
```diff
@@ -0,0 +1,104 @@
+	compatible = "efinix,sapphire";
+
+	chosen {
+
```
done

manoj153 commented on drivers/gpio/gpio_efinix_sapphire.c:
```diff
@@ -0,0 +1,244 @@
+{
+	sapphire_write32(GPIO_OUTPUT_ADDR, value);
+
+	return 0;
```
made it `void `

carlocaione commented:
Can you rebase to re-trigger the CI?

fkokosinski commented on boards/riscv/titanium_ti60_f225/doc/index.rst:
```diff
@@ -0,0 +1,56 @@
+
+The Efinix Titanium Ti60 F225 development kit contains a Ti60 FPGA, which is fabricated on a 16nm process and deliver
+high performance with the lowest possible power on a small physical size. In addition, Efinix offers Sapphire SoC IP,
+which is a user-configurable RISC-V SoC based on the VexRiscv core with configurable feature set and extension.
```
I don't see which Sapphire SoC revision (commit SHA?) this Zephyr platform assumes in this doc and in the linked Sapphire-Zephyr repo. Could you add it somewhere in the documentation? Or is it somewhere in the Sapphire-Zephyr repo, and I've just missed it?

fkokosinski replied:
> Sorry, I don't quite understand "SoC revision (commit SHA?)"? Could you clarify, please?

Since you're using VexRiscv in your SoC, I assumed that your SoC is a soft one. Now, because soft SoC can be changed you risk making this Zephyr port not work. For some soft SoCs in Zephyr, e.g. OpenTitan, we have the SoC [revision specified in the board docs](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/riscv/opentitan_earlgrey/doc/index.rst#building-and-flashing). And by SoC revision I mean the commit SHA from the repository that hosts the SoC.

Another approach is taken by e.g. LiteX VexRiscv platform, which links to an external repository which always generates the same version of the soft SoC: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/riscv/litex_vexriscv/doc/index.rst#litex-vexriscv.

> Efinix Efinity IP Manager (software tool) generates the Sapphire SoC. Do you mean you mean the version of the software tool?

What I meant is that I want to have a description somewhere on how to generate the exact soft SoC that is Zephyr port assumes, immune to future changes. If giving the version of this tool guarantees that, then that's fine by me.

fkokosinski replied:
Thanks!

manoj153 commented:
> Can you rebase to re-trigger the CI?

Sure, done.

fkokosinski commented on soc/riscv/riscv-privilege/efinix-sapphire/soc.h:
```diff
@@ -0,0 +1,47 @@
+
+#ifndef _ASMLANGUAGE
+
+static inline uint32_t sapphire_read32(uint32_t addr)
+{
+	return sys_read32(addr);
+}
```
IMO you should drop these wrappers for now, and re-add them later when their addition will be better justified, i.e. actually needed and used.

manoj153 replied:
I have dropped the wrappers now.

cfriedt commented:
This looks interesting to me particularly because it seems the board & fpga are also supported with Yosys

https://github.com/YosysHQ/yosys/blob/master/techlibs/efinix/synth_efinix.cc

cfriedt commented on boards/riscv/titanium_ti60_f225/Kconfig.board:
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2023 Efinix Inc.
```
@manoj153 - can you verify that you are an Efinix employee? I just want to ensure that copyright is assigned properly here.

manoj153 replied:
Yes

cfriedt commented on drivers/gpio/gpio_efinix_sapphire_priv.h:
```diff
@@ -0,0 +1,19 @@
+/*
```
If this header is not being used elsewhere in Zephyr, can it be merged into the driver?

manoj153 replied:
Alright, I have now combined it with the driver source.

cfriedt commented on drivers/gpio/gpio_efinix_sapphire.c:
```diff
@@ -0,0 +1,229 @@
+/*
```
Please ensure that this (and all of the drivers you have added in this change set) are added to `tests/drivers/build_all`

manoj153 replied:
1. I've included a GPIO test case in this pull request. Could you please review it and let me know if it fulfils your requirements?

2. There isn't a serial directory in tests/drivers/build_all. Would you like me to create one?

cfriedt reviewed (dismissed):
Drivers need to be added to `tests/drivers/build_all`

manoj153 committed:
tests: drivers: build_all: gpio: add efinix_sapphire

Build only test for efinix_sapphire gpio driver.

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

carlocaione commented on soc/riscv/riscv-privilege/efinix-sapphire/Kconfig.soc:
```diff
@@ -0,0 +1,18 @@
+
+config SOC_RISCV32_EFINIX_SAPPHIRE
+	bool "Efinix Sapphire VexRiscv system implementation"
+	select ATOMIC_OPERATIONS_C
```
Are you sure this should not be `ATOMIC_OPERATIONS_BUILTIN` instead?

manoj153 replied:
Thank you for your suggestion. However, several RISC-V privilege SoCs, including the ones listed below, currently utilize ATOMIC_OPERATIONS_C: (I'm **not sure** why!)

```
soc/riscv/litex-vexriscv/Kconfig.soc:   select ATOMIC_OPERATIONS_C
soc/riscv/openisa_rv32m1/Kconfig.soc:   select ATOMIC_OPERATIONS_C
soc/riscv/riscv-privilege/opentitan/Kconfig.soc:        select ATOMIC_OPERATIONS_C
soc/riscv/riscv-privilege/neorv32/Kconfig.soc:  select ATOMIC_OPERATIONS_C
soc/riscv/riscv-privilege/sifive-freedom/Kconfig.soc:   select ATOMIC_OPERATIONS_C
soc/riscv/riscv-privilege/sifive-freedom/Kconfig.soc:   select ATOMIC_OPERATIONS_C
soc/riscv/riscv-privilege/sifive-freedom/Kconfig.soc:   select ATOMIC_OPERATIONS_C
soc/riscv/riscv-privilege/niosv/Kconfig.soc:    select ATOMIC_OPERATIONS_C
soc/riscv/riscv-privilege/gd32vf103/Kconfig.series:     select ATOMIC_OPERATIONS_C
soc/riscv/esp32c3/Kconfig.defconfig:config ATOMIC_OPERATIONS_C
```

I've attempted to implement ATOMIC_OPERATIONS_BUILTIN for the current SoC in the PR but encountered an issue: during startup, it is causing a exception to `_Fault()`

Therefore, for the time being, we might need to continue with `ATOMIC_OPERATIONS_C` unless we have a good answer.

carlocaione replied:
@npitre given #57674 do you have any idea about this?

manoj153 replied:
This is strange I only see the issue when I enable `CONFIG_SHELL=y` along with `ATOMIC_OPERATIONS_BUILTIN`

manoj153 replied:
@carlocaione This is done now, turns out I made an config mistake on the soft SoC. 

Things are working fine now.

carlocaione commented on soc/riscv/riscv-privilege/efinix-sapphire/soc.h:
```diff
@@ -0,0 +1,18 @@
+#include <zephyr/arch/common/sys_io.h>
+#include <zephyr/devicetree.h>
+
+#ifndef _ASMLANGUAGE
+
+#endif /* _ASMLANGUAGE */
```
This file is empty. Is this needed at all?

manoj153 replied:
I left it because the `zephyr/include/zephyr/arch/riscv/irq.h` line  26: has ` #include <soc.h>` (dependency)
so removing the file `soc.h` is causing compilation error.

carlocaione replied:
@gmarull just because you love `soc.h` ;)

carlocaione commented on drivers/serial/uart_efinix_sapphire.c:
```diff
@@ -0,0 +1,89 @@
+#define UART0_PARITY          0 /* Off */
+#define UART0_STOP            0 /* 1 stop bit */
+
+#define BSP_UART_DATA          0x00
+#define BSP_UART_STATUS        0x04
+#define BSP_UART_CLOCK_DIVIDER 0x08
+#define BSP_UART_FRAME_CONFIG  0x0C
```
Move these before they are actually used.

manoj153 replied:
Fixed

carlocaione commented on tests/drivers/build_all/gpio/testcase.yaml:
```diff
@@ -11,3 +11,9 @@ tests:
       - spi
       - i2c
+
+  drivers.gpio.build.efinix_sapphire:
+    min_ram: 32
+    platform_allow: titanium_ti60_f225
+    depends_on: gpio
+    extra_args: DTC_OVERLAY_FILE="efinix_sapphire.overlay"
```
this doesn't look right to me

carlocaione replied:
@cfriedt this is what you meant with that comment?

manoj153 replied:
@cfriedt pinging. Can you let us know?

manoj153 replied:
@cfriedt @carlocaione How can we find a solution? The PR has been stagnant :( For some time

manoj153 replied:
@npitre @galak  would you be able to do a review and +1?

manoj153 replied:
@cfriedt ping. Please review and let me know if any changes is required.

carlocaione commented on soc/riscv/riscv-privilege/efinix-sapphire/Kconfig.series:
```diff
@@ -0,0 +1,9 @@
+config SOC_SERIES_EFINIX_SAPPHIRE
+	bool "Efinix Sapphire SOC implementation"
+	select RISCV
+	select SOC_FAMILY_RISCV_PRIVILEGE
```
`_PRIVILEGED` see https://github.com/zephyrproject-rtos/zephyr/pull/59090
